ISPN-5298 HotRod millisecond precision for lifespan/maxidle

gustavonalle opened the pull request:
https://issues.jboss.org/browse/ISPN-5298

* Bumped Hot Rod protocol version to 2.2
* Support for all time units from the API. Timespan and maxidle are sent using vLong. Timeunits are encoded in an extra byte (4 bits timespan + 4 bits maxIdle)
* Added test for millisecond precision
* Updated documentation
* Removed hardocoded server version on JCache TCK remote tests

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCache.java:
```diff
@@ -124,6 +124,23 @@
+    * @param maxIdleTimeUnit {@link java.util.concurrent.TimeUnit} for maxIdle
+    * @return true if the value was replaced
+    */
+   boolean replaceWithVersion(K key, V newValue, long version, int lifespan, TimeUnit lifespanTimeUnit, int maxIdle, TimeUnit maxIdleTimeUnit);
```
I am just being picky here, but have we limited ourselves in the protocol to only supporting ints and not longs already?  I am just curious because embedded everything is a long.

gustavonalle replied:
ops, those should be long...

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/TimeUnitParam.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class TimeUnitParam {
+
+   private static final EnumMap<TimeUnit, Byte> timeUnitToByte = new EnumMap<>(TimeUnit.class);
```
This can be declared as a regular Map.

gustavonalle replied:
ok

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec10.java:
```diff
@@ -47,6 +47,12 @@ public void writeClientListenerParams(Transport transport, ClientListener client
 
+   @Override
+   public void writeExpirationParams(Transport transport, long lifespan, long maxIdle, byte timeUnits) {
+      transport.writeVInt((int) lifespan);
```
Maybe we should print a warning message when they use a value that is too large to be sent in the old protocols.

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec20.java:
```diff
@@ -56,6 +56,12 @@ public void writeClientListenerParams(Transport transport, ClientListener client
 
+   @Override
+   public void writeExpirationParams(Transport transport, long lifespan, long maxIdle, byte timeUnits) {
+      transport.writeVInt((int) lifespan);
```
Same here regarding warning message.

wburns commented on server/hotrod/src/main/scala/org/infinispan/server/hotrod/CacheDecodeContext.scala:
```diff
@@ -32,7 +31,7 @@ class CacheDecodeContext(server: HotRodServer) extends ServerConstants with Log
 
    val isTrace = isTraceEnabled
-   val SecondsInAMonth = 60 * 60 * 24 * 30
+   val MillisInAMonth = 60 * 60 * 24 * 30 * 1000L
```
I see understand why we call it this, but maybe this should be just MillisIn30Days instead.  When I first saw this I had flash backs to DB2 date diff approximations [1]

[1] http://www-01.ibm.com/support/knowledgecenter/SSEPGG_9.7.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000861.html?cp=SSEPGG_9.7.0%2F2-10-3-2-155

wburns commented on server/hotrod/src/main/scala/org/infinispan/server/hotrod/CacheDecodeContext.scala:
```diff
@@ -261,18 +260,28 @@ class CacheDecodeContext(server: HotRodServer) extends ServerConstants with Log
-   protected def toMillis(lifespan: Int): Long = {
-      if (lifespan > SecondsInAMonth) {
-         val unixTimeExpiry = TimeUnit.SECONDS.toMillis(lifespan) - System.currentTimeMillis
+   protected def toMillis(param: ExpirationParam): Long = {
```
We already had this before, but this method just rubs me the wrong way :(

gustavonalle replied:
Is this rule (Seconds/Milliseconds in a month) really needed or is it because Hot Rod only supported seconds?

wburns replied:
@galderz would know more, this is the first time I have seen this guy and it gives me the chills

galderz replied:
@gustavonalle We did that to only support one unit. The idea came from the Memcached protocol.

gustavonalle replied:
ok, but removing this completely is problematic now, it may break older clients that relies in the UNIX timestamp rule, wdyt?

wburns replied:
For older versions I would say keep it as is, but for 8.0 I would say we remove this, what do you guys think?

gustavonalle replied:
@wburns you mean for Hot Rod server >= 2.2 remove it, but keep the behaviour for < 2.2?

danberindei replied:
+1 to remove it for 2.2, but keep support in the server if the client has an older protocol

gustavonalle replied:
@danberindei yes, at the end of the day, it'd be better not to change behaviour for old protocols versions. I will change it.

gustavonalle replied:
@danberindei Or we could keep it in 2.2 (that needs to be backported unfortunately), and remove for 8.0 that for sure will need Hot Rod protocol change, wdyt?

danberindei replied:
I'd rather modify the client itself to do the 30days check when backporting, does that sound feasible?

gustavonalle replied:
ok, removing 30d check for 2.2 and keeping the behaviour for < 2.2 is not complicated and worst case 2.2 clients relying on 30d behaviour will just need to call cache ops sending the right unit, now that we support all of them

wburns commented on server/hotrod/src/main/scala/org/infinispan/server/hotrod/Decoder2x.scala:
```diff
@@ -153,14 +143,26 @@ object Decoder2x extends AbstractVersionedDecoder with ServerConstants with Log
+         } else duration
+      }
+      version match {
+         case VERSION_22 =>
```
It seems maybe we should make a Decoder22.scala and have this in there instead and just override the method?

gustavonalle replied:
It could be, but time (no pun intended) is a precious resource a this stage :smile:

wburns replied:
Maybe some other day then

galderz commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/MillisecondExpirationTest.java:
```diff
@@ -0,0 +1,32 @@
+   @Test
+   public void testDefaultExpiration() throws Exception {
+      remoteCache.put("Key", "Value", 50, TimeUnit.MILLISECONDS);
+      Thread.sleep(100);
```
We've had problems in the past with expiration tests randomly failing due to slow execution or GC pauses. For more robust expiration test approaches, I'd recommend checking tests like `org.infinispan.client.hotrod.ExpiryTest`

gustavonalle replied:
I'll take a look

gustavonalle replied:
hmm, was looking at ```org.infinispan.client.hotrod.ExpiryTest``` and it seems it always passes since all ```cache.get``` returns null, regardless of expirations. Missing setting the correct ```keyEquivalence``` maybe?

gustavonalle replied:
Apart from that, ```org.infinispan.client.hotrod.ExpiryTest``` also uses ```Thread.sleep```, and ```DefaultExpirationTest``` (which my test inherits from) never failed in CI, so I'm inclined to leave it 'as is' for now...

danberindei commented on documentation/src/main/asciidoc/user_guide/chapter-64-Hot_Rod_Protocol.adoc:
```diff
@@ -1533,6 +1534,37 @@ converter. If custom marker is `2`, it represents the byte array, as returned
++0x03+ = MICROSECONDS +
++0x04+ = MINUTES +
++0x05+ = HOURS +
++0x06+ = DAYS +
```
I'd like to add two special units, `DEFAULT` and `INFINITE`, and skip the values for these two units.

Without `DEFAULT`, a regular `put(k, v)` would use the `DEFAULT_LIFESPAN` and `DEFAULT_MAXIDLE` flags, plus a "default time unit" and a 0 value for each. 

Without `INFINITE`, overriding the server defaults to make a value un-expirable would require sending two `-1` values, and with the `vLong` encoding each would take 10 bytes.

gustavonalle replied:
I assume we'd need to get rid of the flags ```DEFAULT_LIFESPAN``` and ```DEFAULT_MAXIDLE``` since having a ```DEFAULT``` special unit would be enough?

gustavonalle replied:
(talking only about version >= 2.2)

danberindei replied:
Yep, we'd have to remove those.

wburns commented:
Also it seems like all of the hotrod SSL tests fail now. http://ci.infinispan.org/viewLog.html?buildId=26149&buildTypeId=bt9&guest=1

gustavonalle commented:
@wburns I can't reproduce those failures here... CI is showing exact the same (intermittent) failures for test running on IBM JDK 1.8...Not sure if it's related, will dig it further

gustavonalle commented:
PR updated.

* Removed the 30 days rule for protocol versions >= 2.2 
* Added ```DEFAULT``` and ```INFINITE``` special units to avoid sending the values when not needed.

gustavonalle commented:
I have no local test failures, OTOH It looks like something is massively broken on CI, with plenty of ```Unsupported major.minor version 52.0``` being thrown everywhere.
On the bright side, at least  ```hotrod-server``` and ```hotrod-client``` are passing :smile:

galderz commented:
Looks good to me, any other comments @wburns @danberindei ?

gustavonalle commented:
Btw, TCK tests are passing when combining this PR with https://github.com/infinispan/infinispan/pull/3449 including the millisecond expiration.

danberindei commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/Codec22.java:
```diff
@@ -0,0 +1,26 @@
+   public void writeExpirationParams(Transport transport, long lifespan, TimeUnit lifespanTimeUnit, long maxIdle, TimeUnit maxIdleTimeUnit) {
+      byte timeUnits = encodeTimeUnits(lifespan, lifespanTimeUnit, maxIdle, maxIdleTimeUnit);
+      transport.writeByte(timeUnits);
+      if (lifespan > 0) {
```
I'm not sure about this, I believe in JCache it's possible to have lifespan = 0 so that the entry expires immediately.

gustavonalle replied:
It's fine to have lifespan/maxIdle = 0, they just won't be sent across the network. Please take a look at ```RemoteCacheImpl#applyDefaultExpirationFlags```

danberindei commented:
@gustavonalle @galderz Looks good, only one small comment about the `lifespan > 0` check in Codec22.

danberindei commented:
@gustavonalle @galderz On further consideration, I see `0` must have a special meaning in `RemoteCache` because of the way `put(k, v)` is implemented in `RemoteCacheSupport`, so let's integrate this PR as is.

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/protocol/CodecFactory.java:
```diff
@@ -29,6 +30,7 @@
       codecMap.put(PROTOCOL_VERSION_13, CODEC_13);
       codecMap.put(PROTOCOL_VERSION_20, CODEC_20);
       codecMap.put(PROTOCOL_VERSION_21, CODEC_21);
+      codecMap.put(PROTOCOL_VERSION_22, CODEC_22);
```
:+1: Glad you added this in

wburns commented:
Seems fine to me as well.

gustavonalle commented:
The failures on CI are happening on master as well and ```DistWriteSkewAtomicMapAPITest.testConcurrentWritesAndIteration``` seems not related, it failed randomly recently...

danberindei commented:
Ok, integrating...

danberindei commented:
Done, thanks Gustavo!
